Reachable Panic in `set_timeout()` by negative `tv_sec` value

### Describe the bug
There is a reachable panic in `set_timeout()` at kernel/src/time/core/timer.rs:80 when call syscall `select` with negative `tv_sec` value.

https://github.com/asterinas/asterinas/blob/2bda8d17c42652c6c9276a83e88600567b55fea6/kernel/src/time/core/timer.rs#L80

```rust
1126 #[stable(feature = "duration", since = "1.3.0")]
1127 impl Add for Duration {
1128     type Output = Duration;
1129 
1130     #[inline]
1131     fn add(self, rhs: Duration) -> Duration {
1132         self.checked_add(rhs).expect("overflow when adding durations")
1133     }
1134 }
```

### To Reproduce
1. Compile a program which calls `select`: 
```C
#define _GNU_SOURCE
#include <stdio.h>
#include <sys/time.h>
#include <sys/syscall.h>
#include <unistd.h>
#include <errno.h>

int main(int argc, char *argv[]) {
  struct timeval times[2];
  times[0].tv_sec = -1;
  times[0].tv_usec = 10;
  times[1].tv_sec = -1;
  times[1].tv_usec = 10;

  syscall(SYS_select, NULL, NULL, NULL, NULL, times);
  perror("select");

  return 0;
}

```
2. Run the compiled program in Asterinas.

### Expected behavior

Asterinas reports panic and is terminated.

### Environment
- Official docker asterinas/asterinas:0.8.1
- 11th Gen Intel(R) Core(TM) i7-1165G7 @ 2.80GHz
- Asterinas version: main 2bda8d1

### Logs
```
~ # /root/select.c 
panicked at /root/.rustup/toolchains/nightly-2024-06-20-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/time.rs:1132:31:
overflow when adding durations
Printing stack trace:
   1: fn 0xffffffff88860cf0 - pc 0xffffffff88860d5a / registers:

     rax                0x1; rdx                0x8; rcx                0x0; rbx                0x0;
     rsi                0x1; rdi 0xffffffff8883b137; rbp                0x0; rsp 0xffff80000b2a7f60;

   2: fn 0xffffffff88860b00 - pc 0xffffffff88860c50 / registers:

     rax                0x1; rdx                0x8; rcx                0x0; rbx                0x0;
     rsi                0x1; rdi 0xffffffff8883b137; rbp                0x0; rsp 0xffff80000b2a7fc0;

   3: fn 0xffffffff88049000 - pc 0xffffffff8804900a / registers:

     rax                0x1; rdx                0x8; rcx                0x0; rbx                0x0;
     rsi                0x1; rdi 0xffffffff8883b137; rbp                0x0; rsp 0xffff80000b2a8110;

   4: fn 0xffffffff889d2b80 - pc 0xffffffff889d2c02 / registers:

     rax                0x1; rdx                0x8; rcx                0x0; rbx                0x0;
     rsi                0x1; rdi 0xffffffff8883b137; rbp                0x0; rsp 0xffff80000b2a8120;

   5: fn 0xffffffff889d2e10 - pc 0xffffffff889d2ea0 / registers:

     rax                0x1; rdx                0x8; rcx                0x0; rbx                0x0;
     rsi                0x1; rdi 0xffffffff8883b137; rbp                0x0; rsp 0xffff80000b2a81b0;

   6: fn 0xffffffff889bc590 - pc 0xffffffff889bc5b3 / registers:

     rax                0x1; rdx                0x8; rcx                0x0; rbx                0x0;
     rsi                0x1; rdi 0xffffffff8883b137; rbp                0x0; rsp 0xffff80000b2a8250;

   7: fn 0xffffffff8838ea60 - pc 0xffffffff8838eab1 / registers:

     rax                0x1; rdx                0x8; rcx                0x0; rbx                0x0;
     rsi                0x1; rdi 0xffffffff8883b137; rbp                0x0; rsp 0xffff80000b2a8270;

   8: fn 0xffffffff883f9820 - pc 0xffffffff883f9858 / registers:

     rax                0x1; rdx                0x8; rcx                0x0; rbx                0x0;
     rsi                0x1; rdi 0xffffffff8883b137; rbp                0x0; rsp 0xffff80000b2a82c0;

   9: fn 0xffffffff883f5d70 - pc 0xffffffff883f5e29 / registers:

     rax                0x1; rdx                0x8; rcx                0x0; rbx                0x0;
     rsi                0x1; rdi 0xffffffff8883b137; rbp                0x0; rsp 0xffff80000b2a82f0;

  10: fn 0xffffffff8847c600 - pc 0xffffffff8847c835 / registers:

     rax                0x1; rdx                0x8; rcx                0x0; rbx                0x0;
     rsi                0x1; rdi 0xffffffff8883b137; rbp                0x0; rsp 0xffff80000b2a8470;

  11: fn 0xffffffff88477450 - pc 0xffffffff88477657 / registers:

     rax                0x1; rdx                0x8; rcx                0x0; rbx                0x0;
     rsi                0x1; rdi 0xffffffff8883b137; rbp                0x0; rsp 0xffff80000b2a85d0;

  12: fn 0xffffffff88476f10 - pc 0xffffffff88476faa / registers:

     rax                0x1; rdx                0x8; rcx                0x0; rbx                0x0;
     rsi                0x1; rdi 0xffffffff8883b137; rbp                0x0; rsp 0xffff80000b2a86d0;

  13: fn 0xffffffff88175bf0 - pc 0xffffffff88175c52 / registers:

     rax                0x1; rdx                0x8; rcx                0x0; rbx                0x0;
     rsi                0x1; rdi 0xffffffff8883b137; rbp                0x0; rsp 0xffff80000b2a8750;

  14: fn 0xffffffff88175a30 - pc 0xffffffff88175a8e / registers:

     rax                0x1; rdx                0x8; rcx                0x0; rbx                0x0;
     rsi                0x1; rdi 0xffffffff8883b137; rbp                0x0; rsp 0xffff80000b2a87a0;

  15: fn 0xffffffff88119af0 - pc 0xffffffff88119e37 / registers:

     rax                0x1; rdx                0x8; rcx                0x0; rbx                0x0;
     rsi                0x1; rdi 0xffffffff8883b137; rbp                0x0; rsp 0xffff80000b2a8850;

  16: fn 0xffffffff884d43d0 - pc 0xffffffff884d46ee / registers:

     rax                0x1; rdx                0x8; rcx                0x0; rbx                0x0;
     rsi                0x1; rdi 0xffffffff8883b137; rbp                0x0; rsp 0xffff80000b2a8b50;

  17: fn 0xffffffff884d3660 - pc 0xffffffff884d3eee / registers:

     rax                0x1; rdx                0x8; rcx                0x0; rbx                0x0;
     rsi                0x1; rdi 0xffffffff8883b137; rbp                0x0; rsp 0xffff80000b2a8f20;

  18: fn 0xffffffff884d3430 - pc 0xffffffff884d3501 / registers:

     rax                0x1; rdx                0x8; rcx                0x0; rbx                0x0;
     rsi                0x1; rdi 0xffffffff8883b137; rbp                0x0; rsp 0xffff80000b2a9e40;

  19: fn 0xffffffff882fe4e0 - pc 0xffffffff88307573 / registers:

     rax                0x1; rdx                0x8; rcx                0x0; rbx                0x0;
     rsi                0x1; rdi 0xffffffff8883b137; rbp                0x0; rsp 0xffff80000b2a9f70;

  20: fn 0xffffffff88183d20 - pc 0xffffffff88183dae / registers:

     rax                0x1; rdx                0x8; rcx                0x0; rbx                0x0;
     rsi                0x1; rdi 0xffffffff8883b137; rbp                0x0; rsp 0xffff80000b2c0410;

  21: fn 0xffffffff88194640 - pc 0xffffffff881951e6 / registers:

     rax                0x1; rdx                0x8; rcx                0x0; rbx                0x0;
     rsi                0x1; rdi 0xffffffff8883b137; rbp                0x0; rsp 0xffff80000b2c05b0;

  22: fn 0xffffffff88565200 - pc 0xffffffff8856520e / registers:

     rax                0x1; rdx                0x8; rcx                0x0; rbx                0x0;
     rsi                0x1; rdi 0xffffffff8883b137; rbp                0x0; rsp 0xffff80000b2c0f90;

  23: fn 0xffffffff88898900 - pc 0xffffffff88898916 / registers:

     rax                0x1; rdx                0x8; rcx                0x0; rbx                0x0;
     rsi                0x1; rdi 0xffffffff8883b137; rbp                0x0; rsp 0xffff80000b2c0fb0;

  24: fn 0xffffffff888887a0 - pc 0xffffffff88888809 / registers:

     rax                0x1; rdx                0x8; rcx                0x0; rbx                0x0;
     rsi                0x1; rdi 0xffffffff8883b137; rbp                0x0; rsp 0xffff80000b2c0fd0;


     rax                0x1; rdx                0x8; rcx                0x0; rbx                0x0;
     rsi                0x1; rdi 0xffffffff8883b137; rbp                0x0; rsp 0xffff80000b2c1000;

[OSDK] The kernel seems panicked. Parsing stack trace for source lines:
(  1) /root/asterinas/ostd/src/panicking.rs:113
(  2) /root/asterinas/ostd/src/panicking.rs:59
(  3) cgzpxvl9zvceljk9gh76mj70a:?
(  4) ??:?
(  5) /root/.rustup/toolchains/nightly-2024-06-20-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/panicking.rs:271
(  6) ??:?
(  7) /root/.rustup/toolchains/nightly-2024-06-20-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/option.rs:923
(  8) /root/.rustup/toolchains/nightly-2024-06-20-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/time.rs:1133
(  9) /root/asterinas/kernel/src/time/core/timer.rs:80
( 10) /root/asterinas/kernel/src/time/wait.rs:62
( 11) /root/asterinas/kernel/src/process/signal/pause.rs:111
( 12) /root/asterinas/kernel/src/process/signal/pause.rs:59
( 13) /root/asterinas/kernel/src/process/signal/poll.rs:218
( 14) /root/asterinas/kernel/src/process/signal/poll.rs:171
( 15) /root/asterinas/kernel/src/syscall/poll.rs:88
( 16) /root/asterinas/kernel/src/syscall/select.rs:136
( 17) /root/asterinas/kernel/src/syscall/select.rs:67
( 18) /root/asterinas/kernel/src/syscall/select.rs:26
( 19) /root/asterinas/kernel/src/syscall/mod.rs:198
( 20) /root/asterinas/kernel/src/syscall/mod.rs:328
( 21) /root/asterinas/kernel/src/thread/task.rs:69
( 22) /root/.rustup/toolchains/nightly-2024-06-20-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/ops/function.rs:79
( 23) /root/.rustup/toolchains/nightly-2024-06-20-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/alloc/src/boxed.rs:2077
( 24) /root/asterinas/ostd/src/task/mod.rs:175
make: *** [Makefile:173: run] Error 1
```

--- a/kernel/src/syscall/select.rs
+++ b/kernel/src/syscall/select.rs
@@ -19,8 +19,10 @@ pub fn sys_select(
     let timeout = if timeval_addr == 0 {
         None
     } else {
-        let timeval = ctx.get_user_space().read_val::<timeval_t>(timeval_addr)?;
-        Some(Duration::from(timeval))
+        let mut timeval = ctx.get_user_space().read_val::<timeval_t>(timeval_addr)?;
+        timeval.sec += timeval.usec / 1_000_000;
+        timeval.usec %= 1_000_000;
+        Some(Duration::try_from(timeval)?)
     };
 
     do_sys_select(

--- a/kernel/src/syscall/select.rs
+++ b/kernel/src/syscall/select.rs
@@ -19,9 +19,10 @@ pub fn sys_select(
     let timeout = if timeval_addr == 0 {
         None
     } else {
-        let mut timeval = ctx.get_user_space().read_val::<timeval_t>(timeval_addr)?;
-        timeval.sec += timeval.usec / 1_000_000;
-        timeval.usec %= 1_000_000;
+        let timeval = ctx
+            .get_user_space()
+            .read_val::<timeval_t>(timeval_addr)?
+            .normalize();
         Some(Duration::try_from(timeval)?)
     };
 

--- a/kernel/src/syscall/setitimer.rs
+++ b/kernel/src/syscall/setitimer.rs
@@ -34,8 +34,8 @@ pub fn sys_setitimer(
     }
     let user_space = ctx.get_user_space();
     let new_itimerval = user_space.read_val::<itimerval_t>(new_itimerval_addr)?;
-    let interval = Duration::from(new_itimerval.it_interval);
-    let expire_time = Duration::from(new_itimerval.it_value);
+    let interval = Duration::try_from(new_itimerval.it_interval)?;
+    let expire_time = Duration::try_from(new_itimerval.it_value)?;
 
     let process_timer_manager = ctx.process.timer_manager();
     let timer = match ItimerType::try_from(itimer_type)? {

--- a/kernel/src/time/mod.rs
+++ b/kernel/src/time/mod.rs
@@ -86,9 +86,23 @@ impl From<Duration> for timeval_t {
     }
 }
 
-impl From<timeval_t> for Duration {
-    fn from(timeval: timeval_t) -> Self {
-        Duration::new(timeval.sec as u64, (timeval.usec * 1000) as u32)
+impl TryFrom<timeval_t> for Duration {
+    type Error = crate::Error;
+
+    fn try_from(timeval: timeval_t) -> Result<Self> {
+        if timeval.sec < 0 || timeval.usec < 0 {
+            return_errno_with_message!(Errno::EINVAL, "timeval_t cannot be negative");
+        }
+        if timeval.usec > 1_000_000 {
+            // The value of microsecond cannot exceed 10^6,
+            // otherwise the value for seconds should be set.
+            return_errno_with_message!(Errno::EINVAL, "nsec is not normalized");
+        }
+
+        Ok(Duration::new(
+            timeval.sec as u64,
+            (timeval.usec * 1000) as u32,
+        ))
     }
 }
 

--- a/kernel/src/time/mod.rs
+++ b/kernel/src/time/mod.rs
@@ -21,6 +21,10 @@ pub type time_t = i64;
 pub type suseconds_t = i64;
 pub type clock_t = i64;
 
+const NSEC_PER_USEC: i64 = 1_000;
+const USEC_PER_SEC: i64 = 1_000_000;
+const NSEC_PER_SEC: i64 = 1_000_000_000;
+
 pub(super) fn init() {
     system_time::init();
     clocks::init();
@@ -46,7 +50,7 @@ impl From<Duration> for timespec_t {
 impl From<timeval_t> for timespec_t {
     fn from(timeval: timeval_t) -> timespec_t {
         let sec = timeval.sec;
-        let nsec = timeval.usec * 1000;
+        let nsec = timeval.usec * NSEC_PER_USEC;
         debug_assert!(sec >= 0); // nsec >= 0 always holds
         timespec_t { sec, nsec }
     }
@@ -60,7 +64,7 @@ impl TryFrom<timespec_t> for Duration {
             return_errno_with_message!(Errno::EINVAL, "timesepc_t cannot be negative");
         }
 
-        if value.nsec > 1_000_000_000 {
+        if value.nsec > NSEC_PER_SEC {
             // The value of nanoseconds cannot exceed 10^9,
             // otherwise the value for seconds should be set.
             return_errno_with_message!(Errno::EINVAL, "nsec is not normalized");
@@ -75,6 +79,21 @@ impl TryFrom<timespec_t> for Duration {
 pub struct timeval_t {
     pub sec: time_t,
     pub usec: suseconds_t,
+}
+
+impl timeval_t {
+    /// Normalizes time by adding carries from microseconds to seconds.
+    ///
+    /// Some Linux system calls do this before checking the validity (e.g., the [select]
+    /// implementation).
+    ///
+    /// [select]: https://elixir.bootlin.com/linux/v6.10.5/source/fs/select.c#L716
+    pub fn normalize(&self) -> Self {
+        Self {
+            sec: self.sec.wrapping_add(self.usec / USEC_PER_SEC),
+            usec: self.usec % USEC_PER_SEC,
+        }
+    }
 }
 
 impl From<Duration> for timeval_t {
@@ -93,7 +112,7 @@ impl TryFrom<timeval_t> for Duration {
         if timeval.sec < 0 || timeval.usec < 0 {
             return_errno_with_message!(Errno::EINVAL, "timeval_t cannot be negative");
         }
-        if timeval.usec > 1_000_000 {
+        if timeval.usec > USEC_PER_SEC {
             // The value of microsecond cannot exceed 10^6,
             // otherwise the value for seconds should be set.
             return_errno_with_message!(Errno::EINVAL, "nsec is not normalized");
@@ -101,7 +120,7 @@ impl TryFrom<timeval_t> for Duration {
 
         Ok(Duration::new(
             timeval.sec as u64,
-            (timeval.usec * 1000) as u32,
+            (timeval.usec * NSEC_PER_USEC) as u32,
         ))
     }
 }

